edx/frontend-app-library-authoring
# [BD-14] Implement basic authoring

### Description

Implement basic authoring
    
Adds a first version of library authoring functionality, including the ability to add, view, and edit blocks.
    
This is a straight port of equivalent features in [ramshackle](https://github.com/open-craft/ramshackle), and is meant as a basis for iteration in the near future.  Notably:

* The layout and flow is experimental.
* There is little in the way of theming and i18n.
* As with ramshackle itself, editing certain block types does not work.

### Manual Testing

To test this manually in a devstack, proceed as follows:

1. Check out the latest devstack master and `make dev.clone`.

1.  Check out this PR's branch (`arbrandes/authoring-v1`) in `frontend-app-library-authoring`.

1. Launch the master devstack as usual.

1. Launch `frontend-app-library-authoring` via `make dev.up.frontend-app-library-authoring`, since it is not a default service.

1. Checkout out [blockstore](https://github.com/edx/blockstore/), `make dev.up` and `make dev.provision` it.  In `make blockstore-shell`, create an auth token for edx-platform and a new collection, making a note of its UUID:

    ```bash
    echo "from django.contrib.auth import get_user_model; from rest_framework.authtoken.models import Token; User = get_user_model(); edxapp_user, _ = User.objects.get_or_create(username='edxapp'); Token.objects.get_or_create(user=edxapp_user, key='edxapp-insecure-devstack-key')" | ./manage.py shell
    ./manage.py createcollection "Studio Collection"
    Created Collection: 36aefd6a-c39e-4ede-ab48-fe40ddecfb33 (Studio Collection)
    ```

1. Start blockstore with `make easyserver`.

1. Edit `frontend-app-library-authoring/.env.development` and add the UUID you obtained above to BLOCKSTORE_COLLECTION_UUID:

    ```bash
    BLOCKSTORE_COLLECTION_UUID=36aefd6a-c39e-4ede-ab48-fe40ddecfb33
    ```

1. Restart the `frontend-app-library-authoring` container:

    ```bash
    cd devstack
    docker-compose restart frontend-app-library-authoring
    ```

1. In the Studio shell, configure it with blockstore credentials and enable the `ENABLE_LIBRARY_AUTHORING_MICROFRONTEND` feature flag:

    ```bash
    make studio-shell
    echo "BLOCKSTORE_API_AUTH_TOKEN = 'edxapp-insecure-devstack-key'" >> /edx/app/edxapp/edx-platform/cms/envs/private.py
    echo "BLOCKSTORE_API_AUTH_TOKEN = 'edxapp-insecure-devstack-key'" >> /edx/app/edxapp/edx-platform/lms/envs/private.py
    vim /edx/etc/studio.yml
    ---
    FEATURES:
        ENABLE_LIBRARY_AUTHORING_MICROFRONTEND: true
    ```

1. Exit the shell and restart Studio:

    ```bash
    make studio-restart
    ```

1. In the browser, go to http://localhost:18010/admin/organizations/organization/add/, log in as `edx`, and create an organization named (and short-named)`edX`.

1. Finally, create a `studio.library_authoring_mfe` waffle flag via http://localhost:18010/admin/waffle/flag/, and enable it for everyone.

1. Now go to Studio Home (http://localhost:18010/home/) and click on the Libraries tab.

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_21-27-08](https://user-images.githubusercontent.com/759355/87212391-a5f00200-c2f4-11ea-80d9-91b0ff09b435.png)
    </details>

1. You should be taken to the MFE as a logged-in user and presented with an empty page.  (If, on the other hand, there are previously existing libraries, they'll be shown here).

1. Click on `New Library` in the upper right.  You should see the form appear.

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-07-10_21-35-13](https://user-images.githubusercontent.com/759355/87212525-5fe76e00-c2f5-11ea-8d3e-9bd29227f5df.png)
    </details>

1. Create a new complex library by filling out the form with the "edX" organization and clicking `CREATE`.  You should be taken to the library detail page.  Check that in this new library, you can see the "Add New Component" form.

    <details>
    <summary>Screenshot</summary>

    ![Screenshot_2020-08-10_14-48-35](https://user-images.githubusercontent.com/759355/89813965-d7890280-db18-11ea-8563-6cfefab84fec.png)
    </details>

1. Add a new poll block with a "poll1" slug.

    <details>
    <summary>Screenshot</summary>
    
    ![Screenshot_2020-08-10_14-51-27](https://user-images.githubusercontent.com/759355/89814154-17e88080-db19-11ea-8093-8342a3e70674.png)
    </details>

1. After clicking "Add Component", you should see the block on the screen, and the "Publish" button on the right should be enabled.

    <details>
    <summary>Screenshot</summary>
    
    ![Screenshot_2020-08-10_14-52-40](https://user-images.githubusercontent.com/759355/89814395-731a7300-db19-11ea-97ca-5f71dc968fbe.png)
    </details>

1. Click on "Publish", on the right.  The "Unpublished Changes" badge on the block card should disappear, and the "Publish" button should be disabled.

    <details>
    <summary>Screenshot</summary>
    
    ![Screenshot_2020-08-10_14-55-43](https://user-images.githubusercontent.com/759355/89814551-ae1ca680-db19-11ea-9fe0-bff0f6b3e2cf.png)
    </details>

1. Click on the "View" button in the poll block card.  You should be taken to the block view page:

    <details>
    <summary>Screenshot</summary>
    
    ![Screenshot_2020-08-12_21-35-01](https://user-images.githubusercontent.com/759355/90081459-ccce9900-dce3-11ea-82c6-219359d997e7.png)
    </details>

1. Proceed to make changes to the block as you would on ramshackle.

    <details>
    <summary>Screenshot: edit tab</summary>

    ![Screenshot_2020-08-12_21-37-08](https://user-images.githubusercontent.com/759355/90081549-07d0cc80-dce4-11ea-8257-702dee833851.png)
    </details>

    <details>
    <summary>Screenshot: assets tab</summary>
    
    ![Screenshot_2020-08-12_21-37-50](https://user-images.githubusercontent.com/759355/90081594-1dde8d00-dce4-11ea-8f55-c8baa12c459f.png)
    </details>

    <details>
    <summary>Screenshot: source tab</summary>

    ![Screenshot_2020-08-12_21-38-50](https://user-images.githubusercontent.com/759355/90081649-3d75b580-dce4-11ea-9925-4cef1b538eb2.png)
    </details>

    <details>
    <summary>Screenshot: learn tab</summary>

    ![Screenshot_2020-08-12_21-39-23](https://user-images.githubusercontent.com/759355/90081685-567e6680-dce4-11ea-898f-05d452552c78.png)
    </details>
